extensions/kubernetes/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesProcessor.java

```diff
@@ -0,0 +1,151 @@
+package io.quarkus.kubernetes.deployment;
+
+import static io.quarkus.deployment.builditem.ApplicationInfoBuildItem.UNSET_VALUE;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import io.ap4k.Session;
+import io.ap4k.SessionWriter;
+import io.ap4k.docker.annotation.EnableDockerBuild;
+import io.ap4k.docker.generator.DefaultEnableDockerBuildGenerator;
+import io.ap4k.kubernetes.annotation.KubernetesApplication;
+import io.ap4k.kubernetes.generator.DefaultKubernetesApplicationGenerator;
+import io.ap4k.processor.SimpleFileWriter;
+import io.ap4k.project.Project;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationInfoBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
+import io.quarkus.kubernetes.spi.KubernetesPortBuildItem;
+
+class KubernetesProcessor {
+
+    @Inject
+    BuildProducer<GeneratedResourceBuildItem> generatedResourceProducer;
+
+    @Inject
+    BuildProducer<FeatureBuildItem> featureProducer;
+
+    @BuildStep
+    public void build(ApplicationInfoBuildItem applicationInfo,
+            KubernetesConfig kubernetesConfig,
+            List<KubernetesPortBuildItem> kubernetesPortBuildItems) {
+
+        if (kubernetesPortBuildItems.isEmpty()) {
+            return;
+        }
+
+        //this is done to prevent ap4k from failing when the necessary input has not been provided (during tests or devmode for example)
+        if (UNSET_VALUE.equals(applicationInfo.getWiringClassesDir()) || UNSET_VALUE.equals(applicationInfo.getFinalName())) {
+            return;
+        }
+
+        // The resources that ap4k's execution will result in, will later-on be written
+        // by quarkus in the 'wiring-classes' directory
+        // The location is needed in order to properly support s2i build triggering
+        final Path rootPath = Paths.get(applicationInfo.getWiringClassesDir());
+        if (rootPath.toString().isEmpty()) {
+            return;
+        }
+        final Path ap4kRoot = rootPath.resolve("META-INF/ap4k");
+
+        // by passing false to SimpleFileWriter, we ensure that no files are actually written during this phase
+        final SessionWriter sessionWriter = new SimpleFileWriter(ap4kRoot, false);
+        final String packaging = "jar";
+        sessionWriter.setProject(new Project(Paths.get(applicationInfo.getBaseDir()),
+                new io.ap4k.project.BuildInfo(
+                        applicationInfo.getName(), applicationInfo.getVersion(), packaging,
+                        rootPath.getParent().resolve(applicationInfo.getFinalName() + "-runner." + packaging), rootPath)));
```

Comment:
This one seems to assume that the output artifact will be named `-runner` with some extension, what about the native image case?

Reply:
Currently the extension only supports the JVM case (see also the Dockerfile that is being used). We could of course add a configuration property to make it use the native image just as easily.